### PR TITLE
Tools: Add svn-cleanup script

### PR DESCRIPTION
**Pull request by sdixon194**
#### Changes proposed in this Pull Request:

Automates a way to cleanup alphas and betas from previous releases

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

n/a
#### Does this pull request change what data or activity we track or use?

no
#### Testing instructions:

* run `tools/svn-cleanup.sh <plugin>`, although I already used it to cleanup the 11.5 betas/alphas

### CODE REVIEW

File: tools/svn-cleanup.sh [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+JSON=$(curl -s "https://api.wordpress.org/plugins/info/1.0/$WPSLUG.json")
+if ! jq -e '.' <<<"$JSON" &>/dev/null; then
+	die "Failed to retrieve JSON data from https://api.wordpress.org/plugins/info/1.0/$WPSLUG.json"
+fi
+
+# Current stable version
+CURRENT_STABLE_VERSION=$(jq -r .version <<<"$JSON")
+
+# Get all versions, excluding point releases for Jetpack-the-plugin, so we can obtain the previous release cycle's version.
+if [[ $WPSLUG == "jetpack" ]]; then
```
Comment (anomiex) [inline tools/svn-cleanup.sh]: Don't hard-code a slug here. We should probably do the same for any plugin using WordPress-style versioning (i.e. check the plugin's composer.json for `.extra.changlogger.verisoning` being "wordpress".

File: tools/svn-cleanup.sh [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	SVN_TMP=$(jq -r '.versions | keys[] | select( test( "^[0-9]+(\\.[0-9]+){1}$" ) )' <<<"$JSON"  | sort -V )
+else
```
Comment (anomiex) [inline tools/svn-cleanup.sh]: Extraneous space.
```suggestion
else
```

File: tools/svn-cleanup.sh [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+# Checkout and delete the tags, then commit to SVN.
+info "Checking out SVN shallowly to $DIR"
+svn -q checkout "https://plugins.svn.wordpress.org/$WPSLUG/" --depth=empty "$DIR"
+success "Done! Checked out to $DIR"
+
+info "Checking out SVN tags to $DIR/tags (this may take a while)"
+svn -q up tags --depth=empty
+cd tags
+SVN_PREV_TAGS=$(echo "$SVN_PREV_TAGS" | tr '\n' ' ')
+svn -q up ${SVN_PREV_TAGS}
```
Comment (anomiex) [inline tools/svn-cleanup.sh]: Is it really necessary to fully check out the tag in order to remove it? Or could this use `--depth=empty` to get it registered and delete it that way without ever fully fetching all the files inside?

(real question, I'm not too familiar with the ins and outs of svn)

File: tools/svn-cleanup.sh [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+# Check plugin.
+process_plugin_arg "${ARGS[0]}"
+PLUGIN_NAME=$(jq --arg n "${ARGS[0]}" -r '.name // $n' "$PLUGIN_DIR/composer.json")
+WPSLUG=$(jq -r '.extra["wp-plugin-slug"] // ""' "$PLUGIN_DIR/composer.json")
+[[ -n "$WPSLUG" ]] || die "Plugin $PLUGIN_NAME has no WordPress.org plugin slug. Cannot cleanup previous tags."
+
+# Check build dir.
+if [[ -z "$BUILD_DIR" ]]; then
+	TMPDIR="${TMPDIR:-/tmp}"
+	BUILD_DIR=$(mktemp -d "${TMPDIR%/}/update-tag.XXXXXXXX")
```
Comment (anomiex) [inline tools/svn-cleanup.sh]:
```suggestion
	BUILD_DIR=$(mktemp -d "${TMPDIR%/}/svn-cleanup.XXXXXXXX")
```

**Review by anomiex — COMMENTED** [review]
I wonder if we could simplify the logic here. Instead of fetching from the wporg API, could we do something like this
```
svn -q checkout "https://plugins.svn.wordpress.org/$WPSLUG/" --depth=empty "$DIR"
svn -q up tags --depth=immediates
```
then look at the subdirectories in `$DIR/tags/` to figure out what needs deleting?

File: tools/svn-cleanup.sh (hunk 4 above)
Comment (anomiex) [inline tools/svn-cleanup.sh]: You might also want to pick up the change in #27783 here too.

File: tools/svn-cleanup.sh (hunk 3 above)
Comment (sdixon194) [reply tools/svn-cleanup.sh]: That's a good question, I'm not sure. Let me play around and see.

File: tools/svn-cleanup.sh (hunk 1 above)
Comment (sdixon194) [reply tools/svn-cleanup.sh]: Funny that was my first thought before I took the easy way to just try it out. I'll look into reading the plugin's composer.json.

File: tools/svn-cleanup.sh (hunk 3 above)
Comment (sdixon194) [reply tools/svn-cleanup.sh]: Looks like that does indeed work, nice thinking!

**Comment by sdixon194** [comment]
> I wonder if we could simplify the logic here. Instead of fetching from the wporg API, could we do something like this

Hmm we could do this, but I'm thinking we still have to fetch from the wporg API anyway in order to get the current stable tag. Is there another way to grab that consistently? The plugin's composer.json doesn't have it (the autoloader-suffix has the alpha/new cycle), and the readme.txt isn't consistent in the event someone doesn't backport the stable tag update.

**Comment by anomiex** [comment]
You could do `svn up trunk/ --depth=empty` and `svn up trunk/readme.txt`, then read the stable tag from there. That's what the wporg API looks at, as far as I know.

Or just guess that the largest version number tag without a `-` is the current version, and ask the user to confirm. Chances are this will be getting run just after they ran the stable-tag script anyway.

**Comment by sdixon194** [comment]
Okay! this should be good for another review now.

File: tools/svn-cleanup.sh [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+info "Checking out SVN tags to $DIR/tags"
+svn -q up tags --depth=immediates
+cd tags
+success "Done!"
+
+info "Getting list of pre-release tags"
+PRERELEASE_TAGS=()
+for TAG in *
+	do
+		if [[ "$TAG" =~ [0-9]+(\.[0-9]+)+- ]] && version_compare "$STABLE_TAG" "$TAG" && [[ "$TAG" != "$STABLE_TAG"* ]]; then
```
Comment (anomiex) [inline tools/svn-cleanup.sh]: Do we care about the possibility of `"1.10" != "1.1"*` incorrectly not matching? Might be safer to do like
```suggestion
		if [[ "$TAG" =~ [0-9]+(\.[0-9]+)+- ]] && version_compare "$STABLE_TAG" "$TAG" && [[ "$TAG" != "$STABLE_TAG".* && "$TAG" != "$STABLE_TAG"-* ]]; then
```

File: tools/svn-cleanup.sh [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+info "Checking out SVN tags to $DIR/tags"
+svn -q up tags --depth=immediates
+cd tags
+success "Done!"
+
+info "Getting list of pre-release tags"
+PRERELEASE_TAGS=()
+for TAG in *
+	do
```
Comment (anomiex) [inline tools/svn-cleanup.sh]: [nitpick] While this is ok, I usually write these a little more compactly:
```bash
for TAG in *; do
    ...
done
```
In particular, it'll save a weird-looking double-unindent between lines 107 and 108 below.

File: tools/svn-cleanup.sh [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+done
+proceed_p "" "Continue?"
+echo ""
+
+info "Deleting tags..."
+svn -q rm "${PRERELEASE_TAGS[@]}"
+svn ci -m "Deleting previous release's alphas and betas"
+success "Done!"
+
+exit
```
Comment (anomiex) [inline tools/svn-cleanup.sh]: [nitpick] While it doesn't hurt anything, there's no point to an explicit `exit` (or `exit 0`) at the end of the script.

**Review by anomiex — COMMENTED** [review]
Feel free to ignore the comments labeled "nitpick".

File: tools/svn-cleanup.sh (hunk 6 above)
Comment (sdixon194) [reply tools/svn-cleanup.sh]: Ah I love this suggestion, those looked funny to me too.

File: tools/svn-cleanup.sh (hunk 5 above)
Comment (sdixon194) [reply tools/svn-cleanup.sh]: Updated 👍

File: tools/svn-cleanup.sh [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+svn -q checkout "https://plugins.svn.wordpress.org/$WPSLUG/" --depth=empty "$DIR"
+svn -q up trunk/ --depth=empty
+svn -q up trunk/readme.txt
+
+# Ignore point releases for wordpress-style versioning
+if [[ "$VERSION_TYPE" == "wordpress" ]]; then
+	STABLE_TAG=$(grep "Stable tag:" trunk/readme.txt | grep -Eo '[0-9]+\.([0-9]+)')
+else
+	STABLE_TAG=$(grep "Stable tag:" trunk/readme.txt | grep -Eo '[0-9]+(\.[0-9]+)+')
+fi
```
Comment (anomiex) [inline tools/svn-cleanup.sh]: Whitespace nitpick
```suggestion
fi
```

**Review by anomiex — APPROVED** [review]
Let's go for it!

**Pull request merged**